.github/instructions/github-workflows.instructions.md

```diff
@@ -0,0 +1,1087 @@
+- [ ] Matrix strategy for parallel operations
+- [ ] `retention-days` set appropriately for artifacts
+
+### Testing
+
+- [ ] Unit tests run early in pipeline
+- [ ] Integration tests configured with service dependencies
+- [ ] E2E tests run against appropriate environment
+- [ ] Test reports uploaded as artifacts
+- [ ] Tests can run on forks (without secrets where possible)
+
+### Deployment
+
+- [ ] Environment protection configured for sensitive deployments
+- [ ] Manual approvals for production
+- [ ] Rollback strategy documented and implemented
+- [ ] Health checks validate deployments
+- [ ] Fork protection on deployment jobs
+
+### Structure
+
+- [ ] Clear job names representing pipeline phases
+- [ ] `needs` dependencies properly defined
+- [ ] Conditional execution uses `if` appropriately
+- [ ] Step names are descriptive
+- [ ] Concurrency configured to prevent conflicts
+- [ ] Timeout configured for long-running jobs
+
+## Troubleshooting
+
+### Workflow Not Triggering
+
+1. Check `on` triggers match the event
+2. Verify `paths`, `branches` filters are correct
+3. For `workflow_dispatch`, ensure file is in default branch
+4. Review `if` conditions that might skip execution
+5. Check concurrency settings for blocking runs
+
+### Permission Errors
+
+1. Review `permissions` at workflow and job level
+2. Verify secret access and environment configuration
+3. For OIDC, check trust policy in cloud provider
+4. Confirm repository context for fork-safe operations
+
+### Cache Misses
+
+1. Validate cache key uses `hashFiles()` correctly
+2. Ensure `path` matches actual dependency location
+3. Use `restore-keys` for fallback patterns
+4. Check cache size limits
+
+### Flaky Tests
+
+1. Add explicit waits (avoid `sleep`)
+2. Ensure test isolation and cleanup
+3. Use stable selectors for E2E tests
+4. Implement retries for transient failures
+5. Capture screenshots/videos on failure
+
+### Fork Testing Issues
+
+1. Verify `workflow_dispatch` is uncommented for testing
+2. Check repository-specific `if` conditions
+3. Ensure Make targets work without secrets
+4. Test with minimal environment setup
+
+## Conclusion
+
+These guidelines combine industry best practices for GitHub Workflows with Radius-specific patterns that prioritize testability, security, and developer experience. By following these principles, you'll create workflows that are:
+
+- **Testable:** Core logic runs locally via Make, workflows can be tested on forks
+- **Secure:** Least privilege, secret management, OIDC, dependency scanning
+- **Efficient:** Caching, parallelization, shallow clones, smart artifact usage
+- **Maintainable:** DRY principle, reusable workflows, clear naming conventions
+- **Reliable:** Comprehensive testing, environment protection, rollback strategies
+
+Remember: Workflows are code. Apply the same rigor to workflow development as you would to application code.
+
+---
+
+<!-- End of GitHub Workflows Best Practices for Radius -->
+
+### 1. Unit Tests
```

Comment:
This section has a lot of content that would not apply to Radius workflows, like references to test types and test setup techniques. I think most of the principles here about what types of tests to create and what they do would be in the `constitution.md` or the `agents.md` file, or even the golang instructions because they apply to test implementation.

Reply:
removed